dss/api/collections.py

```diff
@@ -0,0 +1,185 @@
+
+    if version is None:
+        # list the collections and find the one that is the most recent.
+        prefix = "collections/{}.".format(uuid)
+        for matching_key in handle.list(bucket, prefix):
+            matching_key = matching_key[len(prefix):]
+            if version is None or matching_key > version:
+                version = matching_key
+    try:
+        collection_blob = handle.get(bucket, "{}/{}.{}".format(COLLECTION_PREFIX, uuid, version))
+    except BlobNotFoundError:
+        raise DSSException(404, "not_found", "Could not find collection for UUID {}".format(uuid))
+    return json.loads(collection_blob)
+
+@dss_handler
+def get(uuid: str, replica: str, version: str = None):
+    authenticated_user_email = request.token_info['email']
+    collection_body = get_impl(uuid=uuid, replica=replica, version=version)
+    if collection_body["owner"] != authenticated_user_email:
+        raise DSSException(requests.codes.forbidden, "forbidden", f"Collection access denied")
+    return collection_body
+
+@dss_handler
+def find(replica: str):
+    raise NotImplementedError()
+
+@dss_handler
+def put(json_request_body: dict, replica: str, uuid: str, version: str):
+    authenticated_user_email = request.token_info["email"]
+    collection_body = dict(json_request_body, owner=authenticated_user_email)
+    uuid = uuid.lower()
+    handle = Config.get_blobstore_handle(Replica[replica])
+    verify_collection(collection_body["contents"], Replica[replica], handle)
+    collection_uuid = uuid if uuid else str(uuid4())
+    if version is not None:
+        # convert it to date-time so we can format exactly as the system requires (with microsecond precision)
+        timestamp = iso8601.parse_date(version)
+    else:
+        timestamp = datetime.datetime.utcnow()
+    collection_version = datetime_to_version_format(timestamp)
+    handle.upload_file_handle(Replica[replica].bucket,
+                              "{}/{}.{}".format(COLLECTION_PREFIX, collection_uuid, collection_version),
+                              io.BytesIO(json.dumps(collection_body).encode("utf-8")))
+    return jsonify(dict(uuid=collection_uuid, version=collection_version)), requests.codes.created
+
+class hashabledict(dict):
+    def __hash__(self):
+        return hash(tuple(sorted(self.items())))
+
+@dss_handler
+def patch(uuid: str, json_request_body: dict, replica: str, version: str):
+    # patch one collection with another
+    uuid = uuid.lower()
+    handle = Config.get_blobstore_handle(Replica[replica])
+    try:
+        cur_collection_blob = handle.get(Replica[replica].bucket, "{}/{}.{}".format(COLLECTION_PREFIX, uuid, version))
+    except BlobNotFoundError:
+        raise DSSException(404, "not_found", "Could not find collection for UUID {}".format(uuid))
+    collection = json.loads(cur_collection_blob)
+    for field in "name", "description", "details":
```

Comment:
so you can patch someone else's collection?

Reply:
Also, you're not validating that the collection was tombstoned.  I think this is probably the right model, as it circumvents some nasty race condition issues, but it also creates some weirdness where someone can revive a collection that was tombstoned.